drivers/gps/nrf9160_gps/nrf9160_gps.c

```diff
@@ -964,36 +655,193 @@ static int init(const struct device *dev, gps_event_handler_t handler)
 
 	err = configure_antenna();
 	if (err) {
+		LOG_ERR("Failed to configure antenna, error: %d", err);
 		return err;
 	}
 
-	if (drv_data->socket < 0) {
-		int ret = open_socket(drv_data);
+	err = nrf_modem_gnss_init();
+	if (err) {
+		LOG_ERR("Failed to initialize GNSS API, error: %d", err);
+		return err;
+	}
 
-		if (ret != 0) {
-			return ret;
-		}
+	err = nrf_modem_gnss_event_handler_set(gnss_event_handler);
+	if (err) {
+		LOG_ERR("Failed to set GNSS event handler, error: %d", err);
+		return err;
 	}
 
+	init_thread(dev);
+
+	k_work_init(&drv_data->start_work, start_work_fn);
 	k_work_init(&drv_data->stop_work, stop_work_fn);
 	k_work_init_delayable(&drv_data->timeout_work, timeout_work_fn);
 	k_work_init_delayable(&drv_data->blocked_work, blocked_work_fn);
-	k_sem_init(&drv_data->thread_run_sem, 0, 1);
 
-	err = init_thread(dev);
+	atomic_set(&drv_data->is_init, 1);
+	atomic_set(&drv_data->is_running, 0);
+
+	return 0;
+}
+
+static int start(const struct device *dev, struct gps_config *cfg)
+{
+	int err;
+	struct gps_drv_data *drv_data = dev->data;
+
+	if (atomic_get(&drv_data->is_init) != 1) {
+		LOG_WRN("GPS must be initialized first");
+		return -ENODEV;
+	}
+
+	if (atomic_get(&drv_data->is_running)) {
+		LOG_DBG("Restarting GPS");
+
+		cancel_works(drv_data);
+
+		atomic_set(&drv_data->is_running, 0);
+
+		err = nrf_modem_gnss_stop();
+		if (err) {
+			LOG_WRN("Stopping GPS failed, error: %d", err);
+			/* Try to restart GPS regardless of the error. */
+		}
+	}
+
+	if (parse_cfg(cfg, &drv_data->config) != 0) {
+		LOG_ERR("Invalid GPS configuration");
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_NRF9160_GPS_HANDLE_MODEM_CONFIGURATION
+	if (enable_gps(dev) != 0) {
+		LOG_ERR("Failed to enable GPS");
+		return -EIO;
+	}
+#endif
+
+	err = nrf_modem_gnss_fix_retry_set(drv_data->config.retry);
 	if (err) {
-		LOG_ERR("Could not initialize GPS thread, error: %d",
-			err);
+		LOG_ERR("Failed to set fix retry value, error: %d", err);
+		return -EIO;
+	}
+
+	err = nrf_modem_gnss_fix_interval_set(drv_data->config.interval);
+	if (err) {
+		LOG_ERR("Failed to set fix interval value, error: %d", err);
+		return -EIO;
+	}
+
+	err = nrf_modem_gnss_nmea_mask_set(drv_data->config.nmea_mask);
+	if (err) {
+		LOG_ERR("Failed to set NMEA mask, error: %d", err);
+		return -EIO;
+	}
+
+	err = nrf_modem_gnss_nmea_mask_set(drv_data->config.nmea_mask);
```

Comment:
Setting of NMEA mask seems to happen twice

Reply:
Oops, removed the extra setting.